cli/src/commands/workspace/add.rs

```diff
@@ -114,19 +119,22 @@ pub fn cmd_workspace_add(
         )?;
     }
 
-    // Copy sparse patterns from workspace where the command was run
     let mut new_workspace_command = command.for_workable_repo(ui, new_workspace, repo)?;
-    let (mut locked_ws, _wc_commit) = new_workspace_command.start_working_copy_mutation()?;
-    let sparse_patterns = old_workspace_command
-        .working_copy()
-        .sparse_patterns()?
-        .to_vec();
-    locked_ws
-        .locked_wc()
-        .set_sparse_patterns(sparse_patterns)
-        .map_err(|err| internal_error_with_message("Failed to set sparse patterns", err))?;
-    let operation_id = locked_ws.locked_wc().old_operation_id().clone();
-    locked_ws.finish(operation_id)?;
+
+    // Copy sparse patterns from the current workspace, unless asked not to.
+    if !args.no_sparse_patterns {
+        let (mut locked_ws, _wc_commit) = new_workspace_command.start_working_copy_mutation()?;
+        let sparse_patterns = old_workspace_command
+            .working_copy()
+            .sparse_patterns()?
+            .to_vec();
+        locked_ws
+            .locked_wc()
+            .set_sparse_patterns(sparse_patterns)
+            .map_err(|err| internal_error_with_message("Failed to set sparse patterns", err))?;
+        let operation_id = locked_ws.locked_wc().old_operation_id().clone();
+        locked_ws.finish(operation_id)?;
+    }
```

Comment:
Nit: use the if expression pattern here, as it makes it in my opinion more readable. 

```rust
  let sparse_patterns = if args.no_sparse_pattern {
    Vec::new()
  } else {
   // old code here
  };
```

Reply:
which would make it a really small diff.

Reply:
So this doesn't work; my testsuite actually fails with this. I think the problem is that the empty sparse pattern set is equivalent to `--clear`, while a reset state i.e. no patterns at all is equivalent to `.`. That's probably a bit too much to change for now.

Reply:
Shouldn't that be `RepoPathBuf::root()`? I think it should work with that.

Reply:
Yes, that's correct.

> while a reset state i.e. no patterns at all

It's actually not "no patterns at all". The full stop you see on line 105 represents the root directory (and implicitly everything under it).

Reply:
I think it's better to keep it the way you have it. The new working copy includes all patterns by default. Seems unnecessary to lock the working copy again just to write the same value to it.

Reply:
To be clear my suggested diff would be: 

```suggestion
    
        let (mut locked_ws, _wc_commit) = new_workspace_command.start_working_copy_mutation()?;
        let sparse_patterns = if args.no_sparse_patterns { 
           // A root pattern is equivalent to not setting a pattern at all. 
           // TODO: improve this?
            vec![RepoPathBuf::root()]
        } else { 
             old_workspace_command
            .working_copy()
            .sparse_patterns()?
            .to_vec();
       };
            locked_ws
            .locked_wc()
            .set_sparse_patterns(sparse_patterns)
            .map_err(|err| internal_error_with_message("Failed to set sparse patterns", err))?;
        let operation_id = locked_ws.locked_wc().old_operation_id().clone();
        locked_ws.finish(operation_id)?;
    }
```

Reply:
Yes. My argument was that if `args.no_sparse_patterns` is true, then lines 126 and 137-142 are unnecessary since they're just setting the sparse patterns to what they already were. Right?

That would change, however, if we always created new working copies without any sparse patterns, which could be a reasonable change to make. Or we could say that the caller always has to specify the sparse patterns when they create the working copy.

Reply:
> Yes. My argument was that if `args.no_sparse_patterns` is true, then lines 126 and 137-142 are unnecessary since they're just setting the sparse patterns to what they already were. Right?

Yes, you're correct and I'm not contesting it, I just wanted to show the full example. 

> That would change, however, if we always created new working copies without any sparse patterns, which could be a reasonable change to make. Or we could say that the caller always has to specify the sparse patterns when they create the working copy.

My suggestion would definitely be better for a later change which you suggested here. It probably is out of scope for this PR though.